File: doc-examples/jdbc-example-records-java/src/main/java/example/Book.java
package example;

import org.jspecify.annotations.Nullable;
import io.micronaut.data.annotation.*;
import java.util.Date;

@MappedEntity
public// <1>
record Book(
        @Id @GeneratedValue @Nullable Long id, // <2>
        @DateCreated @Nullable Date dateCreated,
        String title,
        int pages,
        BookGenre genre) {

    public Book(@Nullable Long id, @Nullable Date dateCreated, String title, int pages) {
        this(id, dateCreated, title, pages, BookGenre.OTHER);
    }
    public Book(String title, int pages, BookGenre genre) {
        this(null, null, title, pages, genre);
    }
}

File: doc-examples/jdbc-example-records-python/src/main/python/example/Book.py
from dataclasses import dataclass
from typing import Annotated

from java.util import Date
from micronaut.data.annotation import DateCreated, GeneratedValue, Id, MappedEntity


@MappedEntity  # <1>
@dataclass(frozen=True)
class Book:
    id: Annotated[int | None, Id, GeneratedValue]  # <2>
    dateCreated: Annotated[Date | None, DateCreated]
    title: str
    pages: int
